### PR TITLE
Fixing the syncing with Mailchimp. Issues of curl requests.

### PR DESCRIPTION
**Pull request by roberto-ebizmarts**
ref #1128

### CODE REVIEW

File: app/code/community/Ebizmarts/MailChimp/Model/Api/ItemSynchronizer.php [hunk 1]
```diff
@@ -287,11 +278,12 @@ public function getMailchimpEcommerceDataTableName()
     }
 
     /**
-     * @return Ebizmarts_MailChimp_Model_Resource_Ecommercesyncdata
+     * @return Ebizmarts_MailChimp_Model_Ecommercesyncdata
      */
     public function getMailchimpEcommerceSyncDataModel()
     {
-        return Mage::getModel('mailchimp/ecommercesyncdata');
+        //return Mage::getModel('mailchimp/ecommercesyncdata');
```
[the remark this answers is not in the document]
Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Model/Api/ItemSynchronizer.php]: Solved

File: app/code/community/Ebizmarts/MailChimp/Model/Api/Orders.php [hunk 2]
```diff
@@ -729,6 +728,12 @@ public function replaceAllOrdersBatch($initialTime, $mailchimpStoreId, $magentoS
             $orderCollection->addFieldToFilter('entity_id', array('gt' => $lastId));
         }
 
+        if(!$this->_ecommerceOrdersCollection){
+            $this->_ecommerceOrdersCollection = $this->createEcommerceProductsCollection();
```
[the remark this answers is not in the document]
Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Model/Api/Orders.php]: Solved.

File: app/code/community/Ebizmarts/MailChimp/Model/Resource/Ecommercesyncdata/Customers/Collection.php [hunk 3]
```diff
@@ -27,14 +27,16 @@ public function _construct()
     /**
      * @param Mage_Customer_Model_Resource_Customer_Collection $preFilteredCustomersCollection
      */
-    public function joinLeftEcommerceSyncData($preFilteredCustomersCollection, $columns = array('m4m.*'))
+    public function joinLeftEcommerceSyncData($preFilteredCustomersCollection, $columns = array())
```
[the remark this answers is not in the document]
Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Model/Resource/Ecommercesyncdata/Customers/Collection.php]: Solved

File: app/code/community/Ebizmarts/MailChimp/Helper/Curl.php [hunk 4]
```diff
@@ -19,26 +19,36 @@ class Ebizmarts_MailChimp_Helper_Curl extends Mage_Core_Helper_Abstract
...
     {
-        $url = isset($options[CURLOPT_URL]) ? $options[CURLOPT_URL] : false;
-        $method = isset($options[CURLOPT_CUSTOMREQUEST]) ? $options[CURLOPT_CUSTOMREQUEST] : false;
-        $userAgent = isset($options[CURLOPT_USERAGENT]) ? $options[CURLOPT_USERAGENT] : Zend_Http_Client::HTTP_1;
-        $headers = isset($options[CURLOPT_HTTPHEADER]) ? $options[CURLOPT_HTTPHEADER] : array();
-        $body = isset($options[CURLOPT_POSTFIELDS]) ? $options[CURLOPT_POSTFIELDS] : '';
-
-        if ($url === false)
+        if ($url === false) {
             return array();
```
[the remark this answers is not in the document]
Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Helper/Curl.php]: Solved

File: app/code/community/Ebizmarts/MailChimp/Model/Api/Orders.php [hunk 5]
```diff
@@ -729,6 +728,12 @@ public function replaceAllOrdersBatch($initialTime, $mailchimpStoreId, $magentoS
             $orderCollection->addFieldToFilter('entity_id', array('gt' => $lastId));
         }
 
+        if(!$this->_ecommerceOrdersCollection){
```
[the remark this answers is not in the document]
Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Model/Api/Orders.php]: Solved. **empty()** function used

File: app/code/community/Ebizmarts/MailChimp/Helper/Curl.php [hunk 6]
```diff
@@ -19,26 +19,36 @@ class Ebizmarts_MailChimp_Helper_Curl extends Mage_Core_Helper_Abstract
...
-        $curl->setOptions($options);
-        $curl->write($method, $url, $userAgent, $headers, $body);
-        $response = $curl->read();
-        $responseBody = $response === false ? '' : Zend_Http_Response::extractBody($response);
-        $curlError = $response === false ? $curl->getError() : '';
-        $info = $curl->getInfo();
-        $curl->close();
+            $curlResult = $curl->getBody();
+        } catch (Exception $e) {
+            Mage::log(__METHOD__ . " " . $e->getTraceAsString(), null, '', true);
```
[the remark this answers is not in the document]
Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Helper/Curl.php]: Solved

File: app/code/community/Ebizmarts/MailChimp/Model/Resource/Ecommercesyncdata/Customers/Collection.php [hunk 7]
```diff
@@ -27,14 +27,16 @@ public function _construct()
...
         $mailchimpTableName = $this->getMailchimpEcommerceDataTableName();
+        $joinCondition      = "m4m.related_id = e.entity_id AND m4m.type = '%s' AND m4m.mailchimp_store_id = '%s'";
         $preFilteredCustomersCollection->getSelect()->joinLeft(
-            array('m4m' => $mailchimpTableName),
-            "m4m.related_id = main_table.entity_id AND m4m.type = '" . Ebizmarts_MailChimp_Model_Config::IS_CUSTOMER
-            . "' AND m4m.mailchimp_store_id = '" . $this->getMailchimpStoreId() . "'",
-            $columns
+            array("m4m" => $mailchimpTableName),
+            sprintf($joinCondition, Ebizmarts_MailChimp_Model_Config::IS_CUSTOMER, $this->getMailchimpStoreId() ),
+            array()
```
Comment (Santiagoebizmarts) [inline app/code/community/Ebizmarts/MailChimp/Model/Resource/Ecommercesyncdata/Customers/Collection.php]: May be better to simply avoid this parameter in stead of sending an empty array.

Comment (roberto-ebizmarts) [reply app/code/community/Ebizmarts/MailChimp/Model/Resource/Ecommercesyncdata/Customers/Collection.php]: Solved.

**Pull request merged**